src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Net.Mime;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            ValidateContent(content);
+            Debug.Assert(content.Headers.ContentType != null);
+            Encoding? sourceEncoding = JsonContent.GetEncoding(content.Headers.ContentType.CharSet);
+
+            return ReadFromJsonAsyncCore(content, type, sourceEncoding, options ?? JsonContent.s_defaultSerializerOptions, cancellationToken);
+        }
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            ValidateContent(content);
+            Debug.Assert(content.Headers.ContentType != null);
+            Encoding? sourceEncoding = JsonContent.GetEncoding(content.Headers.ContentType.CharSet);
+
+            return ReadFromJsonAsyncCore<T>(content, sourceEncoding, options ?? JsonContent.s_defaultSerializerOptions, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, Encoding? sourceEncoding, JsonSerializerOptions? options, CancellationToken cancellationToken)
```

Comment:
Are these `Core` methods called outside of the above two? If not, why not merge them?

Reply:
As per https://github.com/dotnet/runtime/pull/33459#discussion_r393883547 suggests, I did it to make the validations outside of the `async` method.